include/tvm/node/structural_equal.h

```diff
@@ -140,17 +158,58 @@ class SEqualReducer : public BaseValueEqual {
    * \param handler The equal handler for objects.
    * \param map_free_vars Whether or not to map free variables.
    */
-  explicit SEqualReducer(Handler* handler, bool map_free_vars)
-      : handler_(handler), map_free_vars_(map_free_vars) {}
+  explicit SEqualReducer(Handler* handler, const PathTracingData* tracing_data, bool map_free_vars)
+      : handler_(handler), tracing_data_(tracing_data), map_free_vars_(map_free_vars) {}
+
+  /*!
+   * \brief Reduce condition to comparison of two attribute values.
+   * \param lhs The left operand.
+   * \param rhs The right operand.
+   * \return the immediate check result.
+   */
+  bool operator()(const double& lhs, const double& rhs) const;
+  bool operator()(const int64_t& lhs, const int64_t& rhs) const;
+  bool operator()(const uint64_t& lhs, const uint64_t& rhs) const;
+  bool operator()(const int& lhs, const int& rhs) const;
+  bool operator()(const bool& lhs, const bool& rhs) const;
+  bool operator()(const std::string& lhs, const std::string& rhs) const;
+  bool operator()(const DataType& lhs, const DataType& rhs) const;
+
+  template <typename ENum, typename = typename std::enable_if<std::is_enum<ENum>::value>::type>
+  bool operator()(const ENum& lhs, const ENum& rhs) const {
```

Comment:
Every enum is a unique type, so we need a template function for enums. We can't move the entire implementation of this function to `structural_equal.cc` because the client code needs to instantiate the template with different parameters.

To reduce the binary bloat generated by templates (and to hide implementation details from the header), I extracted as much of the common code for this template function into a concrete implementation function `EnumAttrsEqual()` that is placed in `structural_equal.cc`.

Reply:
Thanks for the explanation! I learned a new cool technique today!